### PR TITLE
Cut Posterior Implementation + Examples

### PR DESCRIPTION
**Pull request by agrawalraj**
This PR includes the following new classes to perform modular Bayesian inference:

- `CutModule`: Removes log probability terms from sample sites outside those specified in `vars` 
- `CutComplementModule`: The complement of `CutModule` (i.e., represents the variables in the model that are not in `vars`)
- `IndexCutModule`: Uses plates to represent each module in a single model

To demonstrate the use of modular Bayesian inference for causal inference, we have added two example notebooks that use the classes defined above to implement Bayesian analogues of the synthetic difference-in-difference and DR-learner estimators from https://www.aeaweb.org/articles?id=10.1257/aer.20190159 and https://arxiv.org/abs/2004.14497, respectively.

Pair coded with @eb8680

### CODE REVIEW

**Comment by eb8680** [comment]
The lint and test failures are caused by missing typing functionality in Python 3.8 and could be fixed by replacing `dict` in annotations with `typing.Dict`. I think we should probably also just drop support for Python 3.8.

File: chirho/observational/handlers/soft_conditioning.py [hunk 1]
```diff
@@ -190,3 +203,127 @@ def configure(
         raise NotImplementedError(
             f"Could not reparameterize deterministic site {msg['name']}"
         )
+
+
```
Comment (eb8680) [inline chirho/observational/handlers/soft_conditioning.py:207]: Let's move all of the new code below into a separate file `handlers/cut.py`

**Review by eb8680 — COMMENTED** [review]
Wow, these notebooks look great! It's nice to see the results in both hold up with single-stage estimators via `IndexCutModule`, even if there's still some uncertainty about precisely what distribution it targets.

I left a number of minor comments on the library code, and I'll go over the tests again later today or tomorrow. As far as the notebooks go, I think they're both in a pretty good state, other than some refactoring to simplify the model code in both and the inference code in the SDiD notebook (using `pyro.settings.set(module_local_params=True)` as in the DR-learner notebook should fix the issues with unintended parameter sharing across `Autoguide` instances that you were working around).

I like that their model implementations are structured very similarly to the presentations in their respective papers - I suspect the likely audience of other causal inference experts will find them easier to understand as a result.

I also think that without too much more effort we could now make new, more opinionated versions of one or both examples that really show off the strengths of the overall approach to causal inference we take in ChiRho by explicitly distinguishing between causal models and queries and between estimators for particular queries and further generic sample-splitting-based robustification of those estimators. We can talk more about this sometime later this week.

File: chirho/observational/handlers/soft_conditioning.py (hunk 1 above)
Comment (agrawalraj) [reply chirho/observational/handlers/soft_conditioning.py]: Good idea! I moved this code to cut.py.

File: chirho/observational/handlers/soft_conditioning.py [hunk 2]
```diff
@@ -190,3 +203,127 @@ def configure(
...
+        # 4. The site does not appear in self.vars and is not observed
+        if msg["name"] in self.vars:
+            # use mask to remove the contribution of this observed site to the model log-joint
+            msg["mask"] = (
+                msg["mask"] if msg["mask"] is not None else True
+            ) & torch.tensor(False, dtype=torch.bool).expand(msg["fn"].batch_shape)
+
+    def _pyro_post_sample(self, msg: dict[str, Any]) -> None:
+        if msg["name"] in self.vars:
+            assert msg["is_observed"]
```
[the remark this answers is not in the document]
Comment (agrawalraj) [reply chirho/observational/handlers/soft_conditioning.py]: Yup I agree. I removed this method out.

File: chirho/observational/handlers/soft_conditioning.py [hunk 3]
```diff
@@ -190,3 +203,127 @@ def configure(
...
+
+
+def cut(model: Optional[Callable] = None, *, vars: Set[str] = set()):
+    if model is None:
+        return functools.partial(cut, vars=vars)
+
+    return CutModule(vars)(model), CutComplementModule(vars)(model)
```
[the remark this answers is not in the document]
Comment (agrawalraj) [reply chirho/observational/handlers/soft_conditioning.py]: Done!

File: chirho/observational/handlers/soft_conditioning.py [hunk 4]
```diff
@@ -190,3 +203,127 @@ def configure(
...
+
+
+class IndexCutModule(
```
[the remark this answers is not in the document]
Comment (agrawalraj) [reply chirho/observational/handlers/soft_conditioning.py]: I like `SingleStageCut` too. I updated the name throughout.

File: chirho/observational/handlers/soft_conditioning.py [hunk 5]
```diff
@@ -190,3 +203,127 @@ def configure(
...
+    pyro.poutine.messenger.Messenger
+):  # TODO subclass DependentMaskMessenger
```
[the remark this answers is not in the document]
Comment (agrawalraj) [reply chirho/observational/handlers/soft_conditioning.py]: That makes sense to me.

File: chirho/observational/handlers/soft_conditioning.py [hunk 6]
```diff
@@ -190,3 +203,127 @@ def configure(
...
+        return super().__enter__()
+
+    def _pyro_sample(self, msg: dict[str, Any]) -> None:
+        if pyro.poutine.util.site_is_subsample(msg):
+            return
+        # There are 4 cases to consider for a sample site:
+        # 1. The site appears in self.vars and is observed
+        # 2. The site appears in self.vars and is not observed
+        # 3. The site does not appear in self.vars and is observed
+        # 4. The site does not appear in self.vars and is not observed
```
[the remark this answers is not in the document]
Comment (agrawalraj) [reply chirho/observational/handlers/soft_conditioning.py]: I agree. I removed this comment now.

File: chirho/observational/handlers/soft_conditioning.py [hunk 7]
```diff
@@ -190,3 +203,127 @@ def configure(
...
+
+        # TODO inherit this logic from indexed.handlers.DependentMaskMessenger
+        # use mask to remove the contribution of this observed site to the model log-joint
+
+        cut_index = IndexSet(**{self.name: {0 if msg["name"] in self.vars else 1}})
```
[the remark this answers is not in the document]
Comment (agrawalraj) [reply chirho/observational/handlers/soft_conditioning.py]: Thanks! I used this function with one modification The `get_mask` in DependentMaskMessenger._pyro_sample expects a return type of `torch.Tensor` so I return `indexset_as_mask(IndexSet(**{self.name: {0 if name in self.vars else 1}}))` instead

File: chirho/observational/handlers/soft_conditioning.py [hunk 8]
```diff
@@ -190,3 +203,127 @@ def configure(
...
+        mask = indexset_as_mask(cut_index)  # TODO device
+        msg["mask"] = mask if msg["mask"] is None else msg["mask"] & mask
+
+        # expand distribution to make sure two copies of a variable are sampled
+        msg["fn"] = msg["fn"].expand(
+            torch.broadcast_shapes(msg["fn"].batch_shape, mask.shape)
+        )
```
[the remark this answers is not in the document]
Comment (agrawalraj) [reply chirho/observational/handlers/soft_conditioning.py]: Done!

File: chirho/observational/handlers/soft_conditioning.py [hunk 9]
```diff
@@ -190,3 +203,127 @@ def configure(
...
+
+    def _pyro_post_sample(self, msg: dict[str, Any]) -> None:
+        if pyro.poutine.util.site_is_subsample(msg):
+            return
+
+        if (not msg["is_observed"]) and (msg["name"] in self.vars):
+            # discard the second value
+            # TODO is the mask computed in _pyro_sample correct?
```
[the remark this answers is not in the document]
Comment (agrawalraj) [reply chirho/observational/handlers/soft_conditioning.py]: I agree.

File: tests/observational/test_cut_posterior_modules.py [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+    assert replayed_model_x[1].squeeze() == dummy_guide_x[0].squeeze()
+
+    # Check composability with trace
+    # TODO: ask Eli if the test below should fail based on trace implementation. If not, then
+    # the lines below should be uncommented.
+    # with pyro.poutine.trace() as tr:
+    #     with IndexPlatesMessenger(), IndexCutModule(["x"]):
+    #         pyro.sample("x", dist.Normal(0.0, 1.0))
+
+    # assert tr.trace.nodes["x"]["value"][0] == tr.trace.nodes["x"]["value"][1]
```
[the remark this answers is not in the document]
Comment (agrawalraj) [reply tests/observational/test_cut_posterior_modules.py]: Got it that makes a lot of sense, thank you! I've also added your example as a test case in `test_correctly_duplicates_module_one_vars`.

**Comment by agrawalraj** [comment]
> Wow, these notebooks look great! It's nice to see the results in both hold up with single-stage estimators via `IndexCutModule`, even if there's still some uncertainty about precisely what distribution it targets.
> 
> I left a number of minor comments on the library code, and I'll go over the tests again later today or tomorrow. As far as the notebooks go, I think they're both in a pretty good state, other than some refactoring to simplify the model code in both and the inference code in the SDiD notebook (using `pyro.settings.set(module_local_params=True)` as in the DR-learner notebook should fix the issues with unintended parameter sharing across `Autoguide` instances that you were working around).
> 
> I like that their model implementations are structured very similarly to the presentations in their respective papers - I suspect the likely audience of other causal inference experts will find them easier to understand as a result.
> 
> I also think that without too much more effort we could now make new, more opinionated versions of one or both examples that really show off the strengths of the overall approach to causal inference we take in ChiRho by explicitly distinguishing between causal models and queries and between estimators for particular queries and further generic sample-splitting-based robustification of those estimators. We can talk more about this sometime later this week.

@eb8680 
Thank you, and also for all of the helpful comments! I agree that it would be really nice to show the overall approach we take in ChiRho too. 

For the SDiD notebook, I've been working through several issues with adding `pyro.settings.set(module_local_params=True)`. It messes up inference for the two stage VI procedure (because there's a nested guide defined) but not the single-stage VI procedure. In particular, I get the error `ValueError: optimizer got an empty parameter list`. I've documented a minimal working example showing case this error in the file `test_two_stage_guide.py` based on two different attempts for running SVI (`test_two_stage_inference_attempt1` and `test_two_stage_inference_attempt2`). Do you think you could take a look or find a time to meet to address this issue? Alternatively, we could remove this section because I think we want people to use `SingleStageCut` anyways. Thanks!

**Review by eb8680 — APPROVED** [review]

**Pull request merged**